docs/architecture/adr-004-tips-contract.md

```diff
@@ -0,0 +1,141 @@
+# ADR 004: Tips Contract
+
+## Changelog
+
+- Aug 25, 2022: Initial draft.
+
+## Status
+DRAFTED
+
+## Abstract
+This ADR defines the architecture of the Tips contract. This contract manage sending and tracking of tips made within
+Desmos based applications and exposes queries that allow to gather all users sent and received tips easily.
+
+## Context
+Desmos based applications may want to leverage on the web3 features and one of the most popular one is for sure the possibility to tip users. Currently, it's easy to integrate a feature that allow an application to send tokens as tips between users but very difficult to track those tips in an easy way.
+Plus application creators might want to integrate some form of monetization that allows them to earn a little percentage when user send tips to each other.
+
+## Decision
+The idea here is to write a contract called `Tips`, that handles the sending and tracking of user's tips.
+This contract allows sending tips between users as well tracking sent and received tips for each one of them.
+The tips tracking should be limited to a maximum tracked record number that helps to keep the contract storage as small
+as possible. Also, it doesn't make much sense to store all the tips record that a user received because each one of them will
+most likely check only the latest received.
+Additionally, the contract allows its admin to set a fee that users has to pay in order to be allowed to send a tip. This fee, if not set, is equal to zero.
+
+## Specifications
+Here below the specifications for the contract's messages:
+
+### Messages
+
+#### Instantiate
+```rust
+pub struct InstantiateMsg {
+  pub admin: String,
+  pub subspace_id: u64,
+  pub service_fee: Vec<Coin>,
+  pub saved_tips_record_threshold: u64
+}
+```
+
+* The `admin` identifies the user that controls the contract;
+* The `subspace_id` identifies the application which is deploying the contract;
+* The `service_fee` identifies a fee that the users need to pay to use the contract;
+* The `save_tips_record_threshold` identifies the number of records saved of a user tips history.
+
+#### Execute
+```rust
+pub enum ExecuteMsg{
+  SendTip{post_id: u64, receiver: String},
+  UpdateServiceFee{new_fee: Vec<Coin>},
+  UpdateAdmin{new_admin: String},
+  UpdateSavedTipsRecordThreshold{new_threshold: u64}
+}
+```
+
+##### SendTip
+With the `SendTip` message the user call the contract to send a tip to another user to show their support towards a specific content they made.
+The `MessagInfo` field of the message contains both:
+* the tip identified by the `funds` field
+* the sender identified by the `sender` field
+
+Ideally the message should perform the following checks:
+* Check that the post associated with the given `post_id` exists
+* Check that the `receiver` is equal to the post author
+* If there's any service fee, check that it's covered by the tip amount
```

Comment:
If you want to uniquely identify posts you should also specify the subspace id, as there can be multiple posts with the same ids inside different subspaces. 

That being said, why is tipping here only tied to posts? Why not allowing a generic (non-content related) tip?

Reply:
> If you want to uniquely identify posts you should also specify the subspace id, as there can be multiple posts with the same ids inside different subspaces.

@RiccardoM If you take a look at the `InstantiateMsg{}` it contains also a `subspace_id` field that identifies which subspace the instantiated contract is tied with. This way anyone creating their own social app can decide whether or not to have the tipping feature. So ideally each social can have its own tipping contract (which is identified by an address) that manages tips only for that. `Why` this? 
1. Because relying on only 1 smart contract for all the social networks make the tips managing more complicated and will most likely slow down the execution times of it. 
2. It requires everyone to force  an update of their own client if the contract migrates to a new one because of an update

This is the way with which NFTs contract are deployed. Each collection has its own contract.

> That being said, why is tipping here only tied to posts? Why not allowing a generic (non-content related) tip?

There's no particular reason, I've did it like that because I was looking at how the UI of our product is built and since it doesn't allow any generic tipping I adapted like it. But we can make the `post_id` field optional so we can allow generic tipping with the same message.

Reply:
> If you take a look at the InstantiateMsg{} it contains also a subspace_id field that identifies which subspace the instantiated contract is tied with

All right, missed that. 

> But we can make the post_id field optional so we can allow generic tipping with the same message.

I think it would make more sense to have something more generic, like a `target` field that can be either `PostTarget` or `UserTarget` or (in the future) any other kind of target. This way we can easily distinguish what was the meaning of a tip